### ISSUE TITLE
App gallery doesn't work properly

### ISSUE THREAD

**Issue by ceskyDJ** [ImranR98/Obtainium#1270]
**Prerequisites**

There is a problem with apps from App gallery. I get an error for every update.

**Describe the bug**

When I force scan for updates of apps or automatic scan occurs, I get an error: `FormatException: Trying to read yy from hu-aw-ei at 0 [AI Life]`.

**To Reproduce**

![Screenshot_20240110_003050.jpg](https://github.com/ImranR98/Obtainium/assets/48548230/de8554a1-081f-4699-bd63-7cebca7b3e2a)

**Please complete the following information:**
 - Device: Samsung Galaxy S8
 - OS: Android 14
 - Obtainium Version: 0.15.8-beta

**Comment by ceskyDJ** [comment]
Here is the example of HTTP response headers:
```
curl -D - -sS https://appgallery.cloud.huawei.com/appdl/C10406921
HTTP/1.1 302 Found
Date: Wed, 10 Jan 2024 16:15:13 GMT
Content-Type: application/octet-stream
Content-Length: 0
Connection: keep-alive
Set-Cookie: HWWAFSESID=f72da8860381244ce2; path=/
Set-Cookie: HWWAFSESTIME=1704903304461; path=/
location: https://appdlc-dre.hispace.dbankcloud.cn/dl/appdl/application/apk/2a/2a7dc726066546569d6578231f3f7688/com.huawei.smarthome.2312171752.apk?maple=0&trackId=0&distOpEntity=HWSW
X-Frame-Options: sameorigin
Server: elb
```

I think that `2312171752` (2023-12-17 17:52) is some kind of version. We need to update regex for its extraction.

**Comment by ceskyDJ** [comment]
It looks like these lines handle it: https://github.com/ImranR98/Obtainium/blob/main/lib/app_sources/huaweiappgallery.dart#L39C1-L50C4
```dart
  appIdFromRedirectDlUrl(String redirectDlUrl) {
    var parts = redirectDlUrl
        .split('?')[0]
        .split('/')
        .last
        .split('.')
        .reversed
        .toList();
    parts.removeAt(0);
    parts.removeAt(0);
    return parts.reversed.join('.');
  }
```

I think regex will be better, if there is a support for regexes in Dart (I hope so). You can use something like: `\.([0-9]+)\.apk`.

**Comment by ceskyDJ** [comment]
I don't know Dart, but it could looks something like that:
```dart
  appIdFromRedirectDlUrl(String redirectDlUrl) {
    RegExp exp = RegExp(r'\.([0-9]+)\.apk');
    RegExpMatch? match = exp.firstMatch(redirectDlUrl);

    return match![0];
  }
```

**Comment by ImranR98** [comment]
For some reason I'm getting an incorrect "location" response header when sending the request via Flutter, even though it works fine through curl. Changing the User Agent and other request headers to match the curl request doesn't help. Very similar to https://github.com/ImranR98/Obtainium/issues/854#issuecomment-1705722003.

**Comment by ImranR98** [comment]
For some reason this is now working on my end despite no change in the app (at least no change that I expected to affect this issue) - maybe something on the server side. Could you confirm if it's working for you (or if it's app-specific)?

**Comment by Spp595** [comment]
I have exactly the same problem.  I also can't add another app from appgallery

**Comment by ceskyDJ** [comment]
> For some reason this is now working on my end despite no change in the app (at least no change that I expected to affect this issue) - maybe something on the server side. Could you confirm if it's working for you (or if it's app-specific)?

It looks working now.

**Comment by Spp595** [comment]
sorry for the confusion, updated obtanium

**Comment by ceskyDJ** [comment]
> sorry for the confusion, updated obtanium

So, does it work for you, too?

**Comment by Spp595** [comment]
yes, beta 8 was installed, after updating to the latest everything works fine

**Comment by gregist** [comment]
> For some reason this is now working on my end despite no change in the app (at least no change that I expected to affect this issue) - maybe something on the server side. Could you confirm if it's working for you (or if it's app-specific)?

The update checking for Huawei gallery worked occationally but not consistently. Please reopen this issue.

**Comment by ceskyDJ** [comment]
> > For some reason this is now working on my end despite no change in the app (at least no change that I expected to affect this issue) - maybe something on the server side. Could you confirm if it's working for you (or if it's app-specific)?
> 
> The update checking for Huawei gallery worked occasionally but not consistently. Please reopen this issue.

Ok, so I reopened it.

**Comment by ceskyDJ** [comment]
> The update checking for Huawei gallery worked occasionally but not consistently.

It seems like it works well for me. I haven't got any error for some time (I think since the update, that should not fix it 😅).

**Comment by qqyc** [comment]
> I still have this error on v0.15.11.

Me too.

**Comment by gregist** [comment]
I just noticed that some apps are working properly while some aren't. For example,

huawei health https://appgallery.huawei.com/app/C10414141 updates correctly, while
微信读书 https://appgallery.huawei.com/app/C10398894 throws the error as OP described.

Could it be the Chinese characters in the app name that causes the bug?  Just a guess.  But hope that helps.

**Comment by ceskyDJ** [comment]
> I just noticed that some apps are working properly while some aren't. For example,
> 
> huawei health https://appgallery.huawei.com/app/C10414141 updates correctly, while
> 微信读书 https://appgallery.huawei.com/app/C10398894 throws the error as OP described.
> 
> Could it be the Chinese characters in the app name that causes the bug?  Just a guess.  But hope that helps.

Hmm, it looks like a nice catch! I'll look at it in the evening, how it differs.

**Comment by teteko233** [comment]
hey, I have tested [QQ](https://appgallery.huawei.com/app/C9319). it also doesn't work. So the bug may be not made by Chinese characters  

and, I noticed that I only can add the first party apps developed by appgallery, such like AI life, HMS core. These apps' author are Huawei AppGallery and apk name prefix are com.huawei

**Comment by ceskyDJ** [comment]
> hey, I have tested [QQ](https://appgallery.huawei.com/app/C9319). it also doesn't work. So the bug may be not made by Chinese characters
> 
> and, I noticed that I only can add the first party apps developed by appgallery, such like AI life, HMS core. These apps' author are Huawei AppGallery and apk name prefix are com.huawei

It could be the problem.

Headers obtained by curl for the app QQ:

> HTTP/1.1 302 Found
> Date: Mon, 05 Feb 2024 21:12:03 GMT
> Content-Type: application/octet-stream
> Content-Length: 0
> Connection: keep-alive
> Set-Cookie: HWWAFSESID=52d2c82ab23d575f0e; path=/
> Set-Cookie: HWWAFSESTIME=1707167519783; path=/
> location: https://appdlc-dre.hispace.dbankcloud.cn/dl/appdl/application/apk/1b/1b6f570510cd479a9a1a60d047dc9458/com.tencent.mobileqq.2402021025.apk?maple=0&trackId=0&distOpEntity=HWSW
> X-Frame-Options: sameorigin
> Server: elb

**Comment by ceskyDJ** [comment]
@ImranR98, did you try it for apps from different authors than Huawei (with prefix com.huawei)?

**Comment by ImranR98** [comment]
No, didn't think it would make a difference - I'll try some of the apps mentioned here.

**Comment by ceskyDJ** [comment]
It's interesting, that Huawei apps works but 3rd party apps don't. The piece of code I discussed above should be universal, so it is possibly elsewhere.

**Comment by UjuiUjuMandan** [comment]
@gregist @teteko233 

The situation that Huawei apps are fine but orhers are not may be caused by Huawei's data center allocation.

[KFC app](https://appgallery.huawei.com/app/C10640835) for example.

At my home network, China Unicom
`curl -v https://appgallery.cloud.huawei.com/appdl/C10640835` rerurns an incorrect location url as mentioned before:

```
< location: https://appgallery.cloud.huawei.com
```

But as I sshed to a United States VPS, the same command rerurns as:
```
< location: https://appdlc-dra.hispace.dbankcloud.com/dl/appdl/application/apk/a2/a240a62854a2495bb570c657c624f728/com.yek.android.kfc.activitys.2408021632.apk?maple=0&trackId=0&distOpEntity=HWSW
```

Is Huawei geoblocking Chinese IPs, no. Because the following command works in my home network, too.

```
curl https://appgallery.cloud.huawei.com/appdl/C10640835 --connect-to ::159.138.86.75
```

The IP `159.138.86.75`, located in Signature is resolved on my VPS, but my home network resovled Beijing-based `121.36.118.136` doesn't work both on China and US!

So Huawei's Chinese server has no ability of providing direct APK URL.

This is similar of Google, the Beijing Gu Xiang IPs can only provide limited sets of service like

- www.google.cn
- developers.google.cn

But for other GWS-capable services like `www.youtube.com` a 404 would occur.

Ref: https://github.com/SeaHOH/GotoX/wiki/GServers

**Comment by UjuiUjuMandan** [comment]
@ImranR98 At first I tried to MITM this app, but it seems it doesn't trust certificate signed by PCAPdroid CA even it's in system keychain not users, does Flutter has a separate CA storage?

**Comment by ImranR98** [comment]
@UjuiUjuMandan seems so: https://github.com/ImranR98/Obtainium/issues/1258#issuecomment-1888267714
